src/argilla_sdk/records/_dataset_records.py

```diff
@@ -382,11 +396,106 @@ def _normalize_batch_size(self, batch_size: int, records_length, max_value: int)
 
         return norm_batch_size
 
-    def __validate_vector_names(self, vector_names: Union[List[str], str]) -> None:
+    def _validate_vector_names(self, vector_names: Union[List[str], str]) -> None:
         if not isinstance(vector_names, list):
             vector_names = [vector_names]
         for vector_name in vector_names:
             if isinstance(vector_name, bool):
                 continue
             if vector_name not in self.__dataset.schema:
                 raise ValueError(f"Vector field {vector_name} not found in dataset schema.")
+
+    def _infer_record_from_mapping(
+        self,
+        data: dict,
+        mapping: Optional[Dict[str, str]] = None,
+        user_id: Optional[UUID] = None,
+    ) -> "Record":
+        """Converts a mapped record dictionary to a Record object for use by the add or update methods.
+        Args:
+            dataset: The dataset object to which the record belongs.
+            data: A dictionary representing the record.
+            mapping: A dictionary mapping source data keys to Argilla fields, questions, and ids.
+            user_id: The user id to associate with the record responses.
+        Returns:
+            A Record object.
+        """
+        fields: Dict[str, str] = {}
+        responses: List[Response] = []
+        record_id: Optional[str] = None
+        suggestion_values = defaultdict(dict)
+        vectors: List[Vector] = []
+        metadata: Dict[str, MetadataValue] = {}
+
+        schema = self.__dataset.schema
+
+        for attribute, value in data.items():
+            schema_item = schema.get(attribute)
+            attribute_type = None
+            sub_attribute = None
+
+            # Map source data keys using the mapping
+            if mapping and attribute in mapping:
+                attribute_mapping = mapping.get(attribute)
+                attribute_mapping = attribute_mapping.split(".")
```

Comment:
do the mapping always contain a "."? otherwise we might add some extra logic here?

Reply:
A mapping does not need a ".", but we just index the first value in line 441.

This change was migrated with the method from the `Record` class into `DatasetRecords` because it is the only place it's used and its logic all relates to ingestion. What sort of logic would you like to add? This method could benefit from some refactoring.